### PR TITLE
dataset types (software, workflow, etc.) - initial support

### PR DESCRIPTION
**Pull request by pdurbin**
**What this PR does / why we need it**:

This PR provides initial support for dataset types (part of IQSS/dataverse-pm#307):

- a place for dataset types in the database
- ability to use the types via API on create
- send to DataCite "Software" or "Workflow"
- various APIs (see how to test)

A good entry point for docs at https://dataverse-guide--10694.org.readthedocs.build/en/10694/user/dataset-management.html#dataset-types

This pull request also allows the status of feature flags to be listed via API. See https://dataverse-guide--10694.org.readthedocs.build/en/10694/api/native-api.html#list-all-feature-flags

**Which issue(s) this PR closes**:

- Closes IQSS/dataverse#10517
- Closes IQSS/dataverse#10732
- ~~Used to close but we decided against it IQSS/dataverse#10518~~

**Special notes for your reviewer**:

I followed [Proposal: Supporting Multiple Dataset Types in Dataverse](https://docs.google.com/document/d/16RvGXmaPQK9DGsEEbrrFEu8mjUrN93YY_yh7ZONyMDI/edit?usp=sharing) the best I could but I was also influenced by discussions at tech hours.

The only failing test is Shellspec but it should be fixed by #10682

**Suggestions on how to test this**:

Make sure Jenkins is passing.

~~As of this writing (2024-07-31) Jenkins does not have the dataset types feature flag on. Turning this on would test the new feature.~~ (The feature flag was removed.)

Test all APIs:

- https://dataverse-guide--10694.org.readthedocs.build/en/10694/api/native-api.html#api-create-dataset-with-type
- https://dataverse-guide--10694.org.readthedocs.build/en/10694/developers/dataset-semantic-metadata-api.html#api-semantic-create-dataset-with-type
- https://dataverse-guide--10694.org.readthedocs.build/en/10694/api/native-api.html#import-dataset-with-type
- https://dataverse-guide--10694.org.readthedocs.build/en/10694/api/native-api.html#list-all-feature-flags
- https://dataverse-guide--10694.org.readthedocs.build/en/10694/api/native-api.html#show-feature-flag-status

Test publishing to DataCite to ensure that the correct type is sent.

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

Yes, there is a new "Dataset Type" facet:

Screenshot 2024-07-31 at 2 56 34 PM

Also, when you publish a dataset of type software to DataCite, it will show as such in Fabrica. In the example below, look for "Software" next to the name (pyDataverse):

Screenshot 2024-07-29 at 5 16 31 PM

**Is there a release notes update needed for this change?**:

Yes, included.

**Additional documentation**:

Included, see especially:

- https://dataverse-guide--10694.org.readthedocs.build/en/10694/user/dataset-management.html#dataset-types
- https://dataverse-guide--10694.org.readthedocs.build/en/10694/api/native-api.html#list-all-feature-flags

### CODE REVIEW

**Comment by cmbz** [comment]
2024/07/18 - 6.4 proposal request from @poikilotherm

**Comment by pdurbin** [comment]
At tech hours today, I gave a demo of dataset types as of cfac9dc. Here's a screenshot where you can see a new facet called "Dataset Type" and a couple examples of "software" and "workflow":

Screenshot 2024-07-23 at 4 09 21 PM

I pointed out that I've already completed a good amount of the tasks in the issue at #10517 with the exception of this one:

- Send an appropriate type to DataCite (maybe also other PID providers)

However, sending different info to DataCite should probably wait until we merge the following pull request by @qqmyers, to avoid merge conflicts and extra effort:

- https://github.com/IQSS/dataverse/pull/10615

In addition, based on feedback at tech hours, I plan to focus on the following:

- Remove the ability to import software, etc. using DDI. That way we stop abusing the `dataKind` field.
- Make the baseType field non-nullable and create a Flyway script that populates it with "dataset" for existing datasets.
- All new datasets get a type of "dataset" unless something else is specified at create time.
- remove the `_s` from `datasetType_s` and put it in our Solr schema file.

We also discussed the following ideas, but I don't consider any of these blockers for moving this pull request forward:

- Talk with @jggautier and @dliburd about the possibility of icons for software and workflows, to be shown on search cards in the screenshot above, for example.
- Think more about if dataset types (software, workflow, etc.) should be defined in the database rather than being hard-coded in an enum as they are now.
- In [Proposal: Supporting Multiple Dataset Types in Dataverse][] there is a diagram from @poikilotherm and language around extending types through inheritance. We haven't forgotten about this but for now we're following the "incremental development" section of the proposal (adding a single column). We discussed a base type of "text" and how "articles" and "posters" could extend it.

[Proposal: Supporting Multiple Dataset Types in Dataverse]: https://docs.google.com/document/d/16RvGXmaPQK9DGsEEbrrFEu8mjUrN93YY_yh7ZONyMDI/edit?usp=sharing

I'll go ahead and mention members (I can find) of the old [Software, Workflows & Containers Working Group](https://swc.gdcc.io) (not already mentioned above) in case they'd like to see this update on software datasets: @atrisovic @doigl @kmika11 @4tikhonov

**Comment by jggautier** [comment]
@pdurbin, a while back we saw a Dataverse installation that customized those search cards to show the word "Dataset" when the item was a dataset. I wonder if that might be a better approach. Less work to figure out which icons to use for each type of object and maybe clearer for folks searching for data. Although then we have to think about internationalization.

World Agroforestry at https://data.worldagroforestry.org does this. And I think I've seen another installation that does this. Can't find it but it was a bit different visually.

**Review by qqmyers — CHANGES_REQUESTED** [review]
Looks good overall to address the initial issue. I flagged one issue with the json-ld mapping that I think needs to be changed (and will then simplify the code and allow adding the datasetType to the OREMap file more easily). 

Otherwise, I made several comments related to the interaction of the feature flag state, whether the default value has to exist, whether other entries should be added via flyway etc. where I think tweaking the code might help consistency/readability. For these, I don't think there's one right answer, so any response to change/not change is fine - I just wanted to raise the issue.

FWIW: W.r.t. larger design issues: Some of the discussion about whether datasets should have a persistent link to a type (and what happens if a type is changed or added/dropped from those allowed in a collection, etc.) are just premature for this PR - types are currently just an id/name. So even if those concerns lead to design changes later, I don't think this PR should get delayed to try and work through them - probably easier to figure out when additional functionality is getting added.

File: doc/sphinx-guides/source/api/native-api.rst
Comment (pdurbin) [reply doc/sphinx-guides/source/api/native-api.rst:3056]: Hmm, what about another direction.

With collections you can pass "root" or "1" and they both work. Should I switch to that model and get rid of the "byName" stuff?

File: doc/sphinx-guides/source/_static/api/dataset-create-software.jsonld
Comment (pdurbin) [reply doc/sphinx-guides/source/_static/api/dataset-create-software.jsonld:15]: Thanks, I wasn't sure where to go with this.

https://schema.org/Dataset falls under https://schema.org/CreativeWork . Should we go with that? (There's no workflow under CreativeWork, though.) 🤔

File: src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractCreateDatasetCommand.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractCreateDatasetCommand.java:124]: I wondered about that. Yes, it's currently possible to shoot yourself in the foot by deleting the "dataset" type. Sure, I could put something in to prevent this.

File: src/main/java/edu/harvard/iq/dataverse/search/SearchServiceBean.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/search/SearchServiceBean.java:779]: Here's the longish commit message where I tried to explain what's going on: https://github.com/IQSS/dataverse/pull/10694/commits/067d41654b52c486d303c590cfcd483224847feb

Does that help?

File: src/main/java/edu/harvard/iq/dataverse/util/json/JSONLDUtil.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/util/json/JSONLDUtil.java:102]: Great, thanks. I'll circle back to this after we sort out the URI.

File: src/main/java/edu/harvard/iq/dataverse/util/json/JsonPrinter.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/util/json/JsonPrinter.java:410]: Hmm, probably not. It might be a holdover from earlier code where the type was nullable.

File: doc/sphinx-guides/source/_static/api/dataset-create-software.jsonld
Comment (qqmyers) [reply doc/sphinx-guides/source/_static/api/dataset-create-software.jsonld:15]: This URI would be for the 'hasType' relationship/verb rather than the individual values. So perhaps https://schema.org/additionalType. Reading that, it looks like it allows free text types (recommending CamelCase) so would ~work given Software, Dataset, Workflow, etc. Perhaps Croissant folks or others have an opinion.

File: src/main/resources/db/migration/V6.3.0.2.sql
Comment (qqmyers) [reply src/main/resources/db/migration/V6.3.0.2.sql:3]: Minimally I'm just making a comment, but yes I might drop the flag and would suggest considering that. Feature flags are good for just on/off with no persistent state/no other settings, but there's added complexity when you have other dynamic settings/APIs to change things. If you think the feature flag is useful to keep in this case, I'd just suggest going through the other parts of the code and asking whether things like ifs should be if(feature flag) or if(more than one type), etc. to make things as clear as possible.

File: src/main/java/edu/harvard/iq/dataverse/search/SearchServiceBean.java
Comment (qqmyers) [reply src/main/java/edu/harvard/iq/dataverse/search/SearchServiceBean.java:779]: My question/comment is whether the output for a French site is going to have "labels":[{"Le Data":1} ... and it won't just be capitalization in i18n cases.

Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/search/SearchServiceBean.java:779]: My understanding of the code (it was like that when I got there), is that yes, I expect the French version ("Le Data" or whatever is right) to come out of the API.

File: src/main/resources/db/migration/V6.3.0.2.sql
Comment (pdurbin) [reply src/main/resources/db/migration/V6.3.0.2.sql:3]: @qqmyers @scolapasta thanks for the discussion at standup today. As promised, I pushed a branch where I have removed the feature flag:

https://github.com/IQSS/dataverse/compare/10517-dataset-types...10517-dataset-types-no-ff

Please take a look and advise if you'd like me to proceed with this direction. Thanks.

Comment (qqmyers) [reply src/main/resources/db/migration/V6.3.0.2.sql:3]: I think that looks reasonable. The one question I have is whether the code will automatically suppress a facet when there's only one value (or will people who don't add other types still see a new DataType facet that will always show "Dataset" only?) I guess if it isn't suppressed, I might try to add a check to see if there's more than one type defined before adding the facet to search.

Comment (qqmyers) [reply src/main/resources/db/migration/V6.3.0.2.sql:3]: Makes sense to me

Comment (pdurbin) [reply src/main/resources/db/migration/V6.3.0.2.sql:3]: Great. Thanks. I merged the "no feature flag" branch back into the main one for this PR.

File: src/main/resources/db/migration/V6.3.0.2.sql
Comment (pdurbin) [inline src/main/resources/db/migration/V6.3.0.2.sql:4]: In a previous version of this SQL script, I was adding "software" and "workflow" but @scolapasta and @qqmyers think we should avoid this.

Please see 200a45a261 for how I adjusted this SQL script as well as code, tests, and docs that assumed that "software" and workflow would be present.

File: src/main/java/edu/harvard/iq/dataverse/dataset/DatasetType.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/dataset/DatasetType.java:53]: Fixed in f8e8c4f7ee.

File: doc/sphinx-guides/source/api/native-api.rst
Comment (pdurbin) [reply doc/sphinx-guides/source/api/native-api.rst:3060]: @qqmyers and I agreed to support id or name. Done in 867f548cea.

File: src/main/java/edu/harvard/iq/dataverse/api/Datasets.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/api/Datasets.java:5140]: Good catch. I deleted both in 2fb4fa6dca. I don't think we need them.

File: src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractCreateDatasetCommand.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractCreateDatasetCommand.java:124]: Fixed in eb20155779.

File: src/main/java/edu/harvard/iq/dataverse/pidproviders/doi/XmlMetadataTemplate.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/pidproviders/doi/XmlMetadataTemplate.java:213]: I think what we have now, the three hard-coded types, is fine for this initial effort. For a future pull request, maybe we should consider adding a column to the `datasettype` table called `datacitetype` and move the mappings there. I'm open to this or other ideas.

File: src/main/java/edu/harvard/iq/dataverse/util/json/JsonParser.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/util/json/JsonParser.java:339]: I deleted unneeded null checks: 1a834de0e7

File: src/main/java/edu/harvard/iq/dataverse/search/SearchServiceBean.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/search/SearchServiceBean.java:647]: I deleted unneeded null checks: 1a834de0e7

File: src/main/java/edu/harvard/iq/dataverse/util/json/JsonPrinter.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/util/json/JsonPrinter.java:410]: I deleted unneeded null checks: 1a834de0e7

File: src/main/java/edu/harvard/iq/dataverse/util/json/JSONLDUtil.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/util/json/JSONLDUtil.java:101]: I'm using a proper JSON-LD URL as of 19e3e51d7a.

File: doc/sphinx-guides/source/_static/api/dataset-create-software.jsonld
Comment (pdurbin) [reply doc/sphinx-guides/source/_static/api/dataset-create-software.jsonld:15]: I'm using a proper JSON-LD URL as of 19e3e51d7a.

File: src/main/resources/db/migration/V6.3.0.2.sql
Comment (pdurbin) [reply src/main/resources/db/migration/V6.3.0.2.sql:4]: I mentioned this in standup and there were no objections from @qqmyers or @scolapasta. Resolving.

**Comment by pdurbin** [comment]
As of 67e9971 I believe I'd addressed all the code review so far. I'm moving this into ready for review. @qqmyers please clear your request changes if you think we can go forward now.

File: conf/solr/schema.xml
Comment (qqmyers) [inline conf/solr/schema.xml:208]: does this need a copy field entry?

File: doc/release-notes/10517-datasetType.md
Comment (qqmyers) [inline doc/release-notes/10517-datasetType.md:10]: With the line already in the schema.xml above, this should probably evolve into the standard instructions to update schema.xml, or, with custom blocks, to run the update script.  Adding the one (or two) lines to the existing schema.xml is an option (assuming there are no other schema.xml changes n v6.4) but I don't think we usually call that out.

File: doc/sphinx-guides/source/api/native-api.rst
Comment (qqmyers) [inline doc/sphinx-guides/source/api/native-api.rst:3119]: Add that the default "dataset" type can't be deleted.

File: doc/sphinx-guides/source/user/dataset-management.rst
Comment (qqmyers) [inline doc/sphinx-guides/source/user/dataset-management.rst:811]: No longer true

Comment (scolapasta) [reply doc/sphinx-guides/source/user/dataset-management.rst:811]: What if we added an "enabled" flagto the table? That way we could have all 3 in this PR and just leave dataset enabled.

File: src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractCreateDatasetCommand.java
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractCreateDatasetCommand.java:124]: Minor - this is only used in the else clause, so could be defined there.

File: src/main/java/edu/harvard/iq/dataverse/search/SolrSearchResult.java
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/search/SolrSearchResult.java:29]: Time to do the todo

**Review by qqmyers — CHANGES_REQUESTED** [review]
Looks good. I'll assign to @pdurbin to address the minor doc/code changes. Should be ready for QA after that.

File: conf/solr/schema.xml
Comment (pdurbin) [reply conf/solr/schema.xml:208]: Good idea. Added in faace91b20.

File: doc/release-notes/10517-datasetType.md
Comment (pdurbin) [reply doc/release-notes/10517-datasetType.md:10]: I made the instructions more generic in 260ac3acd8.

Yes, the normal boilerplate should be used. This snippet is just a placeholder.

File: doc/sphinx-guides/source/user/dataset-management.rst
Comment (pdurbin) [reply doc/sphinx-guides/source/user/dataset-management.rst:811]: Yes, there's definitely a parallel here with licenses.

I simply deleted the "Supported Dataset Types" section from the docs in 647121acbd. Higher up in the doc we already say `Out of the box, all datasets have a dataset type of "dataset". Superusers can add additional types such as "software" or "workflow"` so the easiest path forward is to simply remove the information that's no longer accurate (I should have removed that section in 200a45a).

As for an `enabled` flag, sure maybe we can add it in a future PR.

File: src/main/java/edu/harvard/iq/dataverse/search/SolrSearchResult.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/search/SolrSearchResult.java:29]: Yeah? Ok, done! See d865521af6

File: src/main/java/edu/harvard/iq/dataverse/util/json/JsonParser.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/util/json/JsonParser.java:335]: Good catch. Simplified in 9c44b3066f.

**Comment by pdurbin** [comment]
@qqmyers I believe I addressed everything. If you agree, please clear your "request changes" status so this PR can be merged (after QA, of course).

I'm moving this into "ready for review".

File: src/main/java/edu/harvard/iq/dataverse/api/Datasets.java
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/api/Datasets.java:5139]: Minor: Should the code also check that name is not numeric (which would break GET as that triggers looking for Id)?

**Review by qqmyers — APPROVED** [review]
Looks good.  I added a couple minor comments.

File: src/main/java/edu/harvard/iq/dataverse/api/Datasets.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/api/Datasets.java:5139]: Good idea. Fixed in 04f1c7cb5b.

File: src/main/resources/db/migration/V6.3.0.2.sql
Comment (pdurbin) [reply src/main/resources/db/migration/V6.3.0.2.sql:4]: Well, the script should only be run once, by Flyway, on deployment. So I think we should be ok.

I did poke around the code base a bit. We don't seem to look for [check_violation](https://www.postgresql.org/docs/current/errcodes-appendix.html) anywhere else.

Comment (qqmyers) [reply src/main/resources/db/migration/V6.3.0.2.sql:4]: FWIW: Looks like we use things like https://github.com/IQSS/dataverse/blob/a6b54984039ba8c28d00196aa62a05d1f4c012aa/src/main/resources/db/migration/V5.9.0.1__7440-configurable-license-list.sql#L11-L15

Comment (pdurbin) [reply src/main/resources/db/migration/V6.3.0.2.sql:4]: Sure, the above make sense when we're talking about a uniqueness violation. Thanks. Interesting. Useful for that case.

**Comment by sekmiller** [comment]
It doesn't seem to me that a Feature Flag is being used for this feature. What am I missing?

**Comment by qqmyers** [comment]
It no longer is, but it did originally so some convenience methods that were added to see flags are still here. That code now code be in a separate PR, but seemed small enough to just review/let through in this.

**Comment by sekmiller** [comment]
Minor thing: when you click on the Dataset type facet you get the name from the table probably - not the bundle, so I'm seeing the lower case name - and you wouldn't see any translation provided.
Screen Shot 2024-08-30 at 10 12 20 AM

**Comment by sekmiller** [comment]
Also advocating putting some kind of indicator of Dataset Type in the UI - tag on the card and/or dataset page.

**Comment by sekmiller** [comment]
There are a couple of newly added unused imports in dataverses.java

**Comment by pdurbin** [comment]
As discussed with @sekmiller I did add some tests in 673d775 to assert that the API properly returns the value from the bundle such as "Software" (capital S) rather than the value from the database such as "software" (lower case).

I suspect the facets not showing the correct values could have something to do with this issue:

- #10158

My hope is that after merging the fix for that issue...

- #10158

... the JSF facets will show the correct values, the ones from the bundle.

We also talked a bit about how a visual indicator that you are looking at a dataset would be nice, but we'll defer this to a future pull request. (This was also mentioned in a [previous comment](https://github.com/IQSS/dataverse/pull/10694#issuecomment-2246312304), how we'd like to pull in Julian and Dwayne.)

Finally, we should probably simply close this issue:

- #10518

I'll check with @cmbz about this at standup. @scolapasta and @qqmyers decided we didn't need this after all, so I removed it.

**Comment by pdurbin** [comment]
> My hope is that after merging the fix for that issue...
> 
> - https://github.com/IQSS/dataverse/pull/10158

From a quick test (making a test branch, merging that PR into this one), I'm hopeful that it will be a good fix. I'm now seeing the correct values from the bundle in the facets.

Screenshot 2024-09-04 at 10 28 59 AM

**Comment by pdurbin** [comment]
Just to close the loop on the discussion above, I retested now that #10158 has been merged and as we suspected when you click the "Software" facet...

Screenshot 2024-10-04 at 12 12 49 PM

... it now correctly says "Dataset Type: Software" (uppercase) instead of "Dataset Type: software" (lowercase):

Screenshot 2024-10-04 at 12 12 59 PM

**Pull request merged**
